리액트 8기 양승민 7주차 과제 제출

rukeni opened the pull request:
## 반영사항
- 이번 7차는 6주차때 작업했던 내용을 기반으로 진행하고자 6주차때 썼던 구조를 가져왔습니다.
 
## 왜 그렇게 했는지
- 6주차까지 고민했던 내용에서 더 추상화하고 고도화해보고 싶었습니다.
- scratch부터 시작, 존재하는 프로젝트에서 시작하는 경험도 해보았으니, 제가 충분히 고민했던 방향에서 더 많은 고민을 녹여보는 TDD를 해보고 싶었습니다. 

## 테스트
```npm run lint``` 통과
```npm run test``` 4개 케이스 실패

rukeni committed:
@swc/jest 환경 설치

rukeni committed:
Chore: react-rotuer-dom 버전 6으로 업그레이드

rukeni committed:
6주차 코드 적용

hannut91 commented:
좋은 시도 같습니다 ㅎㅎ 이제 진짜 얼마 안남았습니다. 마지막까지 화이팅입니다!

hannut91 commented on src/apps/services/api.js:
```diff
@@ -20,9 +20,9 @@ export async function fetchRestaurants({ regionName, categoryId }) {
-  const url = 'https://eatgo-customer-api.ahastudio.com'
-    + `/restaurants/${restaurantId}`;
+export async function fetchRestaurantById(restaurantId) {
+  const url = 'https://eatgo-customer-api.ahastudio.com/restaurants'
```
계속 URL이 반복되서 사용하고 있는 것 같네요. 중복을 제거해 보는게 좋을 것 같아요

rukeni committed:
.http Request 새로 추가

1. login 요청
2. 리뷰생성 요청

rukeni committed:
로그인을 하면 accessToken을 받아오는지 테스트한다.

rukeni committed:
이메일과 비밀번호를 넘겨주면 서버로부터 accessToken을 받아오도록 구현

rukeni committed:
로그인 페이지에는 'Login'이라는 텍스트가 있는지 테스트한다.

rukeni committed:
로그인 페이지 구현

rukeni committed:
라우터에 /login 페이지 추가한다

rukeni committed:
로그인페이지가 요소들을 테스트한다.

1. Email 라벨이 있는지 확인한다.
2. Password 라벨이 있는지 확인한다.
3. Login 버튼이 있는지 확인한다.

rukeni committed:
Email, Password 텍스트박스, Login버튼을 추가한다.

rukeni committed:
로그인을 할때 액션을 테스트한다.

rukeni committed:
로그인에 필요한 액션을 구현한다.

setAccessToken: accessToken을 넣어주는 액션
setLoginFields: 인풋으로부터 email 또는 password를 받아오는 액션
login: 서버로부터 인증을 받은accessToken을 이용하여 로그인을 시도하는 액션

rukeni committed:
서버로부터 accessToken을 받아오는 authorize 함수를 테스트한다.

rukeni committed:
accessToken 가져오는 함수를 구현

rukeni committed:
로그인에 필요한 리듀서를 테스트한다.

1. setAccessToken: accessToken을 받아와서 state에 들어가는지 테스트한다.

2. setLoginFields: loginFields항목에 이메일과 패스워드가 들어가는지 테스트한다.

3. 기본 state가 추가됨에 따라 mockState 수정

rukeni committed:
로그인을 위한 리듀서 기능 구현

1. setAccessToken : 토큰 삽입

2. setLoginFields : 인풋박스 value 갱신

3. initialState에 loginFields, auth추가

rukeni committed:
Home에서 Login 라우팅 추가

rukeni committed:
Login 페이지 redux 연동

rukeni committed:
Login.test.jsx 테스트코드에 react-redux mock 추가

rukeni committed:
로그인 페이지 기능 구현

1. handleSubmit : 로그인 로직 연동

2. handleChange: 오동작 인수 수정

3. isLoading true일때 핸들러 추가

4. isError true일때 핸들러 추가

rukeni committed:
Login페이지 isLoading, isError 추가로 테스트코드 initialState에도 해당 내용 추가

rukeni committed:
login 액션 오동작하는 부분 수정

1. setLoading, setError key 수정

2. authorize함수에 넘겨줘야하는 인수 추가

rukeni committed:
authroize실패시 'Invalid email or password' 에러메시지를 발생시킨다.

rukeni committed:
authorize api 테스트 통과를 위해 status 임시로 추가한다.

rukeni committed:
axios 라이브러리 설치

rukeni committed:
axios 도입하여 fetch 제거

rukeni commented:
axios로 바꿔보려고 했는데 axios로 하려니까 테스트가 어렵네요... 계속 삽질중..

rukeni committed:
'jest-mock-axios' 설치

rukeni committed:
'axios.js' __mock__ 폴더에 추가

rukeni committed:
api 테스트코드 fetch mock에서 axios mock으로 변경

rukeni committed:
액세스토큰이 있을때, 로그인 상태를 분기해주는 isLogin 추가

rukeni committed:
isLogin추가에 따른 테스트 코드 수정

rukeni committed:
로그인페이지에 isLogin을 넣어서 true이면 로그아웃 버튼만 보이도록 구현

rukeni commented:
> axios로 바꿔보려고 했는데 axios로 하려니까 테스트가 어렵네요... 계속 삽질중..

수동으로 잘 mock을 만들지 못하여 `jest-mock-axios`라는 라이브러리의 힘을 빌려서 해결하였습니다. 🤧

[https://www.npmjs.com/package/jest-mock-axios](https://www.npmjs.com/package/jest-mock-axios)

hannut91 commented on src/apps/services/request.js:
```diff
@@ -0,0 +1,16 @@
+  baseURL: 'https://eatgo-customer-api.ahastudio.com',
+});
+
+export const loginRequest = (email, password) => axios.create({
+  baseURL: 'https://eatgo-login-api.ahastudio.com',
+}).post('/session', {
+  email,
+  password,
+});
```
API마다 axios인스턴스를 만들려고 하셨던 것 같네요.

만약에 이 값이 어떤 환경에 따라서 다르게 실행되어야 한다면 어떻게 할 수 있을까요?

예를들어서 개발환경에서는 개발용 서버를 바라보고, 실제환경에서는 실제서버를 바라보도록 하려면 어떻게 할 수 있을지 고민해 보세요

hannut91 reviewed (commented):
기능 단위로 나누신 것 같은데, API나 서비스들도 도메인에 있어야 하는 것은 아닌지 고민이 되네요

rukeni committed:
env 추가

rukeni committed:
'dotenv'라이브러리 설치

rukeni committed:
'dotenv' devDependency에서 dependency로 이동

rukeni committed:
개발환경일때는 .env.local을 바라보게하고 프로덕션일때는 .env를 바라볼 수 있도록 웹팩에 분기문 추가

hannut91 commented:
`.env`파일은 GitHub에 올라가면 안되니, `.gitignore`에 추가하는게 좋을 것 같습니다

hannut91 commented:
https://www.npmjs.com/package/dotenv-webpack

를 써보는 것도 좋을 것 같네요. 근데 이런 것 없이도 구현해 내는 모습이 인상적이네요 ㅎㅎ

rukeni commented:
> https://www.npmjs.com/package/dotenv-webpack
> 
> 를 써보는 것도 좋을 것 같네요. 근데 이런 것 없이도 구현해 내는 모습이 인상적이네요 ㅎㅎ

라이브러리가 있었군요 ㅇㅅㅇ... 검색하는 습관이 필요한데 😓

rukeni committed:
공통 모듈로부터 region으로 관심사 분리

1. Action함수 분리
2. api 분리
3. Reducer 분리
4. 그에따른 임포트 수정

rukeni committed:
공통 모듈로부터 category 관심사로 기능 분리

1. Action함수 분리
2. api 분리
3. Reducer 분리
4. 그에따른 임포트 수정

rukeni committed:
region, category로 관심사 분리하면서 필요없는 코드 제거

rukeni committed:
loadInitialData()를 loadRegions와 loadCategories로 분리하여 호출

rukeni committed:
공통모듈로부터 restaurant 관심사로 기능 분리

rukeni committed:
공통모듈로부터 auth 관심사로 기능 분리

1. Action함수 분리
2. api 분리
3. 공통모듈 api.js, api.test.js 삭제(기능이 없어져서)
4. Reducer 분리
5. 임포트 수정

rukeni committed:
Header 컴포넌트 위치를 공통 컴포넌트가 위치할 shared폴더로 이동

rukeni committed:
로딩을 따로 분리한다. '로딩중...'을 노출하는지 테스트 한다.

rukeni committed:
Loading 컴포넌트 구현

rukeni committed:
에러메시지를 뱉는 컴포넌트를 분리하여 에러메시지가 노출되는 테스트 한다.

rukeni committed:
Error 컴포넌트 구현

rukeni committed:
컨테이너 마다 로딩과 에러를 Loading, Error컴포넌트로 대체

rukeni committed:
login action 임포트 경로 수정

hannut91 commented on src/apps/store/reducer.test.js:
```diff
@@ -0,0 +1,137 @@
+    });
+  });
+  context('when previous state is undefined', () => {
+    const initialState = {
```
이건 없어도 될 것 같아요 ㅎㅎ 삭제해 주세요

hannut91 commented on src/features/category/Categories.jsx:
```diff
@@ -0,0 +1,17 @@
+    <ul>
+      {categories.map(({ id, name }) => (
+        <li key={id}>
+          <button
+            type="button"
+            onClick={() => onClick(id)}
+          >
+            {name}
+            {selectedCategory && id === selectedCategory.id && '(V)'}
+          </button>
```
선택됨을 표시하는 관심사를 컴포넌트로 분리해도 좋을 것 같네요

hannut91 commented on src/features/restaurant/Detail.jsx:
```diff
@@ -0,0 +1,17 @@
+      </p>
+      <h2>메뉴</h2>
+      <ul>
+        {menuItems.map((menu) => (
+          <li key={menu.id}>{menu.name}</li>
+        ))}
```
여기서 구조 분해 할당을 하면 문제가 생길거에요. 어떻게 해결할지 고민해 보세요

hannut91 commented on src/features/restaurant/restaurantActions.js:
```diff
@@ -0,0 +1,71 @@
+export function loadRestaurants() {
+  return async (dispatch, getState) => {
+    try {
+      const {
+        selectedRegion: region,
+        selectedCategory: category,
+      } = getState();
+
+      if (!region || !category) {
+        return;
+      }
+
+      dispatch(setLoading('restaurants', true));
```
여기서는 예외가 발생하지 않을테니, try catch문에서 빼는 것은 어떨까요

hannut91 commented on src/features/restaurant/restaurantActions.js:
```diff
@@ -0,0 +1,71 @@
+export function loadRestaurantDetail(restaurantId) {
+  return async (dispatch) => {
+    try {
+      if (!restaurantId) {
+        return;
+      }
+
+      dispatch(setLoading('restaurantDetail', true));
```
여기도 마찬가지로 try catch문에서 빼야겠네요

hannut91 commented on src/features/shared/Error.test.jsx:
```diff
@@ -0,0 +1,15 @@
+
+test('Error', () => {
+  const { container } = render((
+    <MemoryRouter>
+      <Error errorMessage="에러발생" />
+    </MemoryRouter>
```
MemoryRouter는 없어도 될 것 같아요

rukeni committed:
버튼이 렌더링 되는 테스트한다.

rukeni committed:
Button 컴포넌트 구현

rukeni committed:
Loading 테스트코드에서 MemoryRouter 제거

rukeni committed:
TextBox가 렌더링 되었는지 테스트한다.

rukeni committed:
TextBox 컴포넌트 구현

rukeni committed:
LoginFormContainer가 데이터를 가져왔을때 렌더링 하는지 테스트

rukeni committed:
LoginFormContainer 구현

rukeni committed:
LoginForm가 Email TextBox와 Password TextBox를 가지고 있는지 테스트한다.

rukeni committed:
LoginForm 구현

rukeni committed:
Logout컴포넌트는 '로그아웃'텍스트의 버튼을 가지는지 테스트

rukeni committed:
Logout 컴포넌트 구현

rukeni committed:
Review에 평점과 리뷰 내용의 TextBox가 있고 리뷰남기기 버튼이 있는지 테스트한다.

rukeni committed:
Review 컴포넌트 구현

rukeni committed:
Review -> ReviewForm으로 이름 수정

rukeni committed:
ReviewFormContainer가 useSelector와 useDispatch로 데이터 주입하면 잘 동작하는지 테스트

rukeni committed:
ReviewFormContainer 구현

rukeni committed:
postReview액션함수가 dispatch후, setReviews를 부르는지 테스트한다.

rukeni committed:
reviewActions 액션 함수 추가

1. setReviewFields: review필드값 변경 액션
2. setReviews: 배열 갱신 액션
3. postReview: 리뷰 생성 비동기 핸들러

rukeni committed:
리뷰가 생성되면 201을 반환하는지 테스트한다.

rukeni committed:
리뷰 생성 요청 구현

rukeni committed:
createReview mock추가

rukeni committed:
reviewReducer가 잘 동작하는지 테스트한다.

1. setReviews: score와 description은 받아서 기존 reviews.data에 추가한다.

2. setReviewFields: name과 value를 받아서 해당 name을 key로 갖는 reviewFields 프로퍼티의 값을 갱신한다.

rukeni committed:
reviewReducer 구현

rukeni committed:
root reducer에 reviewState, reviewReducer 추가

rukeni committed:
reducer 테스트 코드에 reviewReducer값 추가

rukeni committed:
setAccessToken을 실행할때 로그인 안되는 버그 수정

rukeni committed:
로그인 상태이면 ReviewFormContainer가 보여진다.

rukeni committed:
restaurantId를 상위로 받아서 리뷰달때 타겟으로 사용한다.

rukeni committed:
테스트코드에 isLogin과 reviewFields mockState추가

rukeni committed:
테스트 통과를 위해 env파일 수정

rukeni committed:
.gitignore에 env 파일들 추가

rukeni committed:
reviews fixture생성

rukeni committed:
리뷰 리스트를 렌더링하는지 테스트한다

rukeni committed:
Reviews 구현

rukeni committed:
Reviews 컴포넌트 역순정렬하여 새로추가시 맨위로 노출되도록 수정

rukeni committed:
useSelector로부터 리뷰목록을 가져와서 렌더링이 되는지 테스트한다.

rukeni committed:
ReviewsContainer 구현

rukeni committed:
레스토랑 상세 테스트코드에 리뷰관련 state 추가

rukeni committed:
리뷰를 작성하면 로딩이 돌고 새로운 리뷰를 요청하는지 테스트한다.

rukeni committed:
새로운 리뷰 작성후, 새롭게 fetch해오는 loadReviews 구현

rukeni committed:
리뷰를 새로 작성하고 입력창 value가 초기화 되는지 테스트 한다.

rukeni committed:
새 리뷰 작성후 인풋 value 초기화 로직 추가

rukeni committed:
레스토랑 상세정보 가져올때 리뷰 목록도 채울수있도록 수정

rukeni committed:
onSubmit시 preventDefault 설정

rukeni committed:
레스토랑 상세 하단에 리뷰목록 추가

hannut91 commented on src/App.test.jsx:
```diff
@@ -1,73 +1,73 @@
-import {
-  MemoryRouter,
-} from 'react-router-dom';
+import React from 'react';
```
이건 없어도 될 것 같아요. 사용되지 않은 import를 정리할 수 있는 eslint룰을 적용해봐도 좋을 것 같네요

https://github.com/import-js/eslint-plugin-import

hannut91 commented on src/features/review/Reviews.jsx:
```diff
@@ -0,0 +1,23 @@
+    <>
+      <h2>리뷰</h2>
+      <ul>
+        {reviews.sort((a, b) => b.id - a.id).map(({
```
`sort`명령어는 원본 배열을 변경하는 메서드입니다. 따라서 이렇게 사용하면 잘못된 사이드 이펙트가 발생할 수 있습니다.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort

rukeni replied:
앗 중요한 지적 감사합니다. 덕분에 다시 mutable한 메서드가 무엇이 있었는지 다시 정리해볼 기회로 삼을 수 있는 것 같습니다.

rukeni committed:
ReviewForm이 클릭되었을때와 textbox이 change되었을때를 테스트한다.

rukeni committed:
리뷰를 새로 작성하는 post요청을 할때 에러났을 때 테스트 코드를 추가

rukeni committed:
createReview에서 try catch 제거

rukeni committed:
Login 페이지 dispatch mock 추가

rukeni committed:
로그인 페이지에 있는 기능들 각 관심사별로 분리

rukeni committed:
isLogin이 true일때 로그아웃버튼이 보이는지 테스트코드에 추가

rukeni committed:
Logout버튼 컴포넌트로 변경

rukeni committed:
LoginFormContainer가 로딩중일때와 에러가 발생했을때 렌더링 테스트 추가

rukeni committed:
login 액션에서 error가 발생했을때 테스트 코드 추가

rukeni committed:
authroize할때 에서 try catch 제거

rukeni committed:
로그인 버튼 Log In으로 문구 수정

rukeni committed:
Login 문구 Log In으로 테스트 코드 수정

rukeni committed:
로그아웃 버튼 Log out으로 수정

rukeni committed:
로그아웃 버튼 Log out으로 테스트 코드 문구 수정

rukeni committed:
로그아웃 액션함수 추가

rukeni committed:
로그아웃 리듀서 추가

rukeni committed:
로그아웃 액션 테스트 코드 추가

rukeni committed:
로그아웃 버튼 이벤트 핸들러 추가

rukeni committed:
로그아웃 버튼 클릭하는 테스트 코드 추가

rukeni committed:
로컬스토리지에서 토큰을 관리하는지 테스트한다.

rukeni committed:
로컬 스토리지 호출하는 메서드 mocking

rukeni committed:
토큰 관리하는 함수 구현

1. 토큰 셋팅 : saveToken
2. 토큰 가져오기 : loadToken
3. 토큰 제거 : removeToken

rukeni committed:
login을 하면 로컬스토리지 토큰을 저장하고 logout을 하면 로컬스토리지에서 토큰을 제거한다.

rukeni committed:
로컬스토리지에 토큰이 있으면 리뷰를 남길 수 있도록 Form을 보여준다.

rukeni committed:
react 임포트 구문 제거

rukeni committed:
sort함수를 이뮤터블하게 사용할 수 있도록 배열 복사하여 적용

rukeni committed:
'eslint-plugin-import' 라이브러리 추가

rukeni committed:
'eslint-plugin-import' eslintrc에 추가
